src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs

```diff
@@ -7612,5 +7612,477 @@ .maxstack  1
         }
 
         #endregion
+
+        #region "regression tests"
+
+        [Fact, WorkItem(16195, "https://github.com/dotnet/roslyn/issues/16195")]
+        public void TestMatchWithTypeParameter_01()
+        {
+            var source =
+@"using System;
+class Program
+{
+    public static void Main(string[] args)
+    {
+        Console.Write(M1<int>(2));
+        Console.Write(M2<int>(3));
+        Console.Write(M1<int>(1.1));
+        Console.Write(M2<int>(1.1));
+    }
+    public static T M1<T>(ValueType o)
+    {
+        return o is T t ? t : default(T);
+    }
+    public static T M2<T>(ValueType o)
+    {
+        switch (o)
+        {
+            case T t:
+                return t;
+            default:
+                return default(T);
+        }
+    }
+}
+";
+            CreateStandardCompilation(source).VerifyDiagnostics(
```

Comment:
>CreateStandardCompilation(source).VerifyDiagnostics( [](start = 12, length = 52)

Consider explicitly using C# 7 version, otherwise this test will break once we bump the default version.